### PR TITLE
Add/jetpack sale coupon

### PR DESCRIPTION
**Pull request by IanRamosC**
#### Changes proposed in this Pull Request
Context: pbNhbs-16p-p2

* Adds the data layer and redux store for the Jetpack Coupon Sale
* Adds a Sale Banner on cloud.jetpack.com/pricing
* Removes the Introductory pricing banner if there's a sale running
* Changes the discount percentage based on the discount of the sale (ideally this should use the `sale_cost` property of a product, but it's not on the scope for this change)

_Note: I'm a bit uncertain about removing the introductory pricing banner, because it will be shown until we fetch the data from the backend, and if there's a sale running it would cause the frontend to change abruptly when hiding the banner_

Screenshot for reference:
image

##### Unit tests
- For unit tests: `yarn test-client client/state/marketing/test`
##### Functional test
- Follow the steps 3 and 4 in D66708-code, and sandbox the API
- Checkout this PR locally
- Run both `yarn start` and `yarn start-jetpack-cloud-p`. Another option is to only start Calypso Blue through `yarn start` and use the live link for Jetpack Cloud appended with `?calypso_env=development`
- Go to http://jetpack.cloud.localhost:3001/pricing?calypso_env=development
- Make sure a sale banner appear at the top and the countdown timer works
- Confirm the introductory pricing is hidden if the sale is running
- Make sure the prices in the pricing page has a discount equivalent to what is returned by the API call, and this price is also applied at checkout

### CODE REVIEW

**Review by cbauerman — CHANGES_REQUESTED** [review]
With a quick glance through everything looks good. However, the design of the query component means that if a sales coupon is not returned it will continue to slam the endpoint trying to get one:
Screen Shot 2021-09-24 at 4 30 50 PM

We should make sure that there is some handling for errors or empty results. Otherwise it looks pretty good, but I will be sure to take a bit of a closer look first thing next week.

**Comment by IanRamosC** [comment]
> However, the design of the query component means that if a sales coupon is not returned it will continue to slam the endpoint trying to get one:

Great catch. It seems like the action that calls the http handler is being dispatched multiple times. I'll debug how I set up the actions and try to find the issue.

**Comment by IanRamosC** [comment]
> > However, the design of the query component means that if a sales coupon is not returned it will continue to slam the endpoint trying to get one:
> 
> Great catch. It seems like the action that calls the http handler is being dispatched multiple times. I'll debug how I set up the actions and try to find the issue.

It seems like it was an issue with the completely empty response that the API was returning in case we didn't have a coupon - which is not really an error but doesn't have a response, then the action creator seems to be passing the action back to the `registerHandler` and it just dispatches that action again, making the current request always call itself again.
And that happens because the aPI should be returning an empty `[ coupon => null ]` instead of an empty response. I'm pushing a change to the API in a few minutes.

**Comment by IanRamosC** [comment]
@cbauerman, the API change I referred to in the previous comment is already in review, and will be live as soon as it's aceppted. When it's live, the front-end code will stop hammering the API multiple times. Also, I've added a few other minor changes to the frontend code. 

Edit: the API change is now live. :)

File: client/jetpack-cloud/sections/pricing/sale-banner/index.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+	return (
+		<>
+			{ ! isClosed && isBeforeExpiry && (
+				<div className="sale-banner" role="banner" aria-label={ translate( 'Discount Banner' ) }>
+					<div className="sale-banner__content">
+						<div>
+							<b>{ translate( 'End of Summer Sale!' ) }</b>
+							&nbsp;
+							{ translate( 'Get %(discount)d%% off your first year of Jetpack', {
```
Comment (a8ci18n) [inline client/jetpack-cloud/sections/pricing/sale-banner/index.tsx:59]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **1** times:
`translate( 'Get %d%% off your first year of Jetpack' )` *ES Score: 6*

**Review by cbauerman — APPROVED** [review]
Looks good! I initially tested without pulling the newest changes but you already anticipated the need to hide the banner for a `null` coupon 👍 LGTM

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6669008

Thank you @IanRamosC for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6669008).

**Pull request merged**